src/Controller/ViewerController.php

```diff
@@ -103,4 +119,57 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
     return $response;
   }
 
+  /**
+   * Gets a render array for a cool viewer.
+   *
+   * @param \Drupal\media\Entity\Media $media
+   *   The media entity to view / edit.
+   * @param string $wopi_client
+   *   The WOPI client url.
+   * @param bool $can_write
+   *   Whether this is a viewer (false) or an edit (true). Permissions will
+   *   also be checked.
+   * @param array{closebutton: bool} $options
+   *   Options for the renderer. Current values:
+   *     - "closebutton" if "true" will add a close box. (see COOL SDK)
+   *
+   * @return array|array{error: string}
+   *   A stub render element array, or an array with an error on failure.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The key to use by Collabora is empty or not configured.
+   */
+  protected function getViewerRender(MediaInterface $media, string $wopi_client, bool $can_write, $options = NULL) {
+    $cool_settings = $this->config('collabora_online.settings')->get('cool');
+    $wopi_base = $cool_settings['wopi_base'];
+    $allowfullscreen = $cool_settings['allowfullscreen'] ?? FALSE;
+
+    $id = $media->id();
+
+    $access_token = $this->tokenManager->encode(
+      [
+        'fid' => $id,
+        'uid' => $this->currentUser()->id(),
+        'wri' => $can_write,
+      ],
+      $expire_timestamp,
+    );
+
+    $render_array = [
+      '#wopiClient' => $wopi_client,
+      '#wopiSrc' => urlencode($wopi_base . '/cool/wopi/files/' . $id),
+      '#accessToken' => $access_token,
+      // Convert to milliseconds.
+      '#accessTokenTtl' => $expire_timestamp * 1000,
+      '#allowfullscreen' => $allowfullscreen ? 'allowfullscreen' : '',
+    ];
+    if ($options) {
+      if (isset($options['closebutton']) && $options['closebutton'] == 'true') {
```

Comment:
This code is mostly unchanged since it was moved.
I would like to rethink the full render element and the template, instead of tinkering around the edges.
This should be a follow-up issues.

This said, in this issue we can do _some_ cleanup.

Reply:
For now we always add the closebutton, in the future we can rethink it.